File: .yalc/@entur/atb-mobile-client-sdk/challenge/bluetooth/model/bluetooth-model-ios-lib/Sources/bluetooth-model-ios-lib/ChallengeAdvertisingFrameBinaryFormat.swift
//
//  File.swift
//  
//
//  Created by Thomas Skjolberg on 09/06/2022.
//

import Foundation
import OrderedCollections

enum ChallengeAdvertisingFrameBinaryFormatError: Error {
    case rangeError(String)
}

public class ChallengeAdvertisingFrameBinaryFormat {
    
    static let MAX_PAYLOAD_LENGTH = 23
    static let MAX_CORRELATION_ID = 255
    
    static let HEADER_FORMAT = ChallengeAdvertisingFrameHeaderBinaryFormat()
    
    // Legacy bluetooth Advertising Packet max size is 31
    // 2 bytes for Type fields
    // 2 bytes for Company Identifier Code
    // leaving 27 for header + payload
    
    public static func parse(buffer: inout Data) throws -> ChallengeAdvertisingFrame {
        let headerLength = ChallengeAdvertisingFrameBinaryFormat.HEADER_FORMAT.getLength()
        
        if(buffer.count <= headerLength) {
            throw ChallengeAdvertisingFrameBinaryFormatError.rangeError("Expected size > \(headerLength)")
        }
        
        if(buffer.count > headerLength + ChallengeAdvertisingFrameBinaryFormat.MAX_PAYLOAD_LENGTH) {
            throw ChallengeAdvertisingFrameBinaryFormatError.rangeError("Expected header + payload to \(headerLength + ChallengeAdvertisingFrameBinaryFormat.MAX_PAYLOAD_LENGTH) bytes or less")
        }
        
        let header = ChallengeAdvertisingFrameBinaryFormat.HEADER_FORMAT.parse(headerBytes: &buffer);

        return ChallengeAdvertisingFrame(contents: &buffer, header: header)
    }
    
    public static func newInstance(payload: inout Data, offset: Int, count: Int, clientId: UInt32, frameNumber : UInt32, lastFrame : Bool, correlationId: UInt32, channel : UInt32) throws -> ChallengeAdvertisingFrame {
        
        var buffer = Data(capacity: 27) // max bluetooth frame payload
        
        let header = ChallengeAdvertisingFrameHeader(clientId: clientId, frameNumber: frameNumber, lastFrame: lastFrame, correlationId: correlationId, channel: channel)
        
        try ChallengeAdvertisingFrameBinaryFormat.HEADER_FORMAT.serialize(buffer: &buffer, header: header)

        payload.withUnsafeBytes { buf in
            buffer.append(buf.bindMemory(to: UInt8.self).baseAddress!.advanced(by: offset), count: count)
        }
        
        return ChallengeAdvertisingFrame(contents: &buffer, header:header)
    }
    
    /**
     *
     * Parse with some sanity checking
     *
     * @param content bytes to parse
     * @return a frame, or null if checks did not pass
     */

    
    public static func parseOrNull(content: inout Data) throws -> ChallengeAdvertisingFrame? {
        if(isValid(content: &content)) {
            return try parse(buffer: &content)
        }
        
        return nil
    }
    
    public static func isValid(content: inout Data) -> Bool {
        let headerLength = ChallengeAdvertisingFrameBinaryFormat.HEADER_FORMAT.getLength()
        return content.count > headerLength && content.count <= headerLength + ChallengeAdvertisingFrameBinaryFormat.MAX_PAYLOAD_LENGTH
    }
    
    public static func fromPayload(frames: inout Array<ChallengeAdvertisingFrame>, bytes: inout Data, clientId: Int, correlationId: Int, channel: Int) throws {
        let boundedCorrelationId = UInt32(correlationId % ChallengeAdvertisingFrameBinaryFormat.MAX_CORRELATION_ID)
        
        let count = bytes.count
        var numFrames = count / ChallengeAdvertisingFrameBinaryFormat.MAX_PAYLOAD_LENGTH
        if (count % ChallengeAdvertisingFrameBinaryFormat.MAX_PAYLOAD_LENGTH > 0) {
           numFrames += 1
        }
        
        for i in 0 ..< numFrames {
            let copyStart = i * ChallengeAdvertisingFrameBinaryFormat.MAX_PAYLOAD_LENGTH
            let copyLength = min(count - copyStart, ChallengeAdvertisingFrameBinaryFormat.MAX_PAYLOAD_LENGTH)

            let frame = try newInstance(payload: &bytes, offset:copyStart, count:copyLength, clientId: UInt32(clientId), frameNumber: UInt32(i), lastFrame: (numFrames - 1 == i), correlationId: boundedCorrelationId, channel: UInt32(channel));
            
            frames.append(frame)
        }
    }

    public static func toPayload(frames: inout Array<ChallengeAdvertisingFrame>, count: Int) -> Data {
        var merge = Data()
        let headerLength = ChallengeAdvertisingFrameBinaryFormat.HEADER_FORMAT.getLength()
        for i in 0..<count {
            let frame = frames[i];
            merge.append(contentsOf: frame.contents[headerLength...])
        }
        return merge
    }

    public static func toPayload(frames: inout Array<ChallengeAdvertisingFrame?>, count: Int) -> Data {
        var merge = Data()
        let headerLength = ChallengeAdvertisingFrameBinaryFormat.HEADER_FORMAT.getLength()
        for i in 0..<count {
            let frame = frames[i];
            merge.append(contentsOf: frame!.contents[headerLength...])
        }
        return merge
    }
    
    
}

File: .yalc/@entur/atb-mobile-client-sdk/challenge/bluetooth/model/bluetooth-model-java-lib/src/main/java/no/entur/abt/android/bluetooth/ChallengeAdvertisingFrameBinaryFormat.java
package no.entur.abt.android.bluetooth;

import java.util.ArrayList;
import java.util.List;

import no.entur.abt.android.bluetooth.header.ChallengeAdvertisingFrameHeader;
import no.entur.abt.android.bluetooth.header.ChallengeAdvertisingFrameHeaderBinaryFormat;

public class ChallengeAdvertisingFrameBinaryFormat {

	public static final int MAX_FRAME_PAYLOAD_LENGTH = 23;
	public static final int CORRELATION_ID_MAXVALUE = 255;

	protected static ChallengeAdvertisingFrameHeaderBinaryFormat HEADER_FORMAT = new ChallengeAdvertisingFrameHeaderBinaryFormat();

	// Legacy bluetooth Advertising Packet max size is 31
	// 2 bytes for Type fields
	// 2 bytes for Company Identifier Code
	// leaving 27 for header + payload

	public static ChallengeAdvertisingFrame parse(byte[] buffer) {
		int headerLength = HEADER_FORMAT.getLength();

		if (buffer == null || buffer.length <= headerLength) {
			throw new IllegalArgumentException("Expected size > " + headerLength);
		}
		if (buffer.length > headerLength + MAX_FRAME_PAYLOAD_LENGTH) {
			throw new IllegalArgumentException("Expected header + payload to be " + (headerLength + MAX_FRAME_PAYLOAD_LENGTH) + " bytes or less");
		}
		ChallengeAdvertisingFrameHeader header = HEADER_FORMAT.parse(buffer);

		return new ChallengeAdvertisingFrame(buffer, header);
	}

	protected static ChallengeAdvertisingFrame newInstance(byte[] payload, int clientId, int frameNumber, boolean lastFrame, int correlationId, int channel) {
		ChallengeAdvertisingFrameHeader header = new ChallengeAdvertisingFrameHeader(clientId, frameNumber, lastFrame, correlationId, channel);

		byte[] buffer = new byte[payload.length + HEADER_FORMAT.getLength()];
		HEADER_FORMAT.serialize(buffer, header);
		System.arraycopy(payload, 0, buffer, HEADER_FORMAT.getLength(), payload.length);

		return new ChallengeAdvertisingFrame(buffer, header);
	}

	/**
	 *
	 * Parse with some sanity checking
	 *
	 * @param content bytes to parse
	 * @return a frame, or null if checks did not pass
	 */

	public static ChallengeAdvertisingFrame parseOrNull(byte[] content) {
		if (isValid(content)) {
			return parse(content);
		} else {
			return null;
		}
	}

	public static boolean isValid(byte[] content) {
		return content.length > HEADER_FORMAT.getLength() && content.length <= MAX_FRAME_PAYLOAD_LENGTH + HEADER_FORMAT.getLength();
	}

	public static List<ChallengeAdvertisingFrame> fromPayload(byte[] bytes, int clientId, int correlationId, int channel) {
		int boundedCorrelationId = correlationId % CORRELATION_ID_MAXVALUE;

		List<ChallengeAdvertisingFrame> frames = new ArrayList<>();

		int numFrames = (bytes.length / MAX_FRAME_PAYLOAD_LENGTH);
		if (bytes.length % MAX_FRAME_PAYLOAD_LENGTH > 0) {
			numFrames++;
		}
		for (int i = 0; i < numFrames; i++) {
			int copyStart = i * MAX_FRAME_PAYLOAD_LENGTH;
			int copyLength = Math.min(bytes.length - copyStart, MAX_FRAME_PAYLOAD_LENGTH);
			byte[] framePayload = new byte[copyLength];
			System.arraycopy(bytes, copyStart, framePayload, 0, copyLength);

			frames.add(newInstance(framePayload, clientId, i, numFrames - 1 == i, boundedCorrelationId, channel));
		}

		return frames;
	}

	public static byte[] toPayload(List<ChallengeAdvertisingFrame> frames) {
		return toPayload(frames.toArray(new ChallengeAdvertisingFrame[frames.size()]), frames.size());
	}

	public static byte[] toPayload(ChallengeAdvertisingFrame[] frames, int count) {
		int resultLength = 0;
		for (int i = 0; i < count; i++) {
			ChallengeAdvertisingFrame frame = frames[i];
			resultLength += frame.getContents().length - HEADER_FORMAT.getLength();
		}
		byte[] merge = new byte[resultLength];
		int offset = 0;
		for (int i = 0; i < count; i++) {
			ChallengeAdvertisingFrame frame = frames[i];
			int length = frame.getContents().length - HEADER_FORMAT.getLength();
			System.arraycopy(frame.getContents(), HEADER_FORMAT.getLength(), merge, offset, length);
			offset += length;
		}
		return merge;
	}

}
